processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,343 @@
+			return fmt.Errorf("couldn't add spans to existing trace: %w", err)
+		}
+
+		// we are done with this trace, move on
+		return nil
+	}
+
+	// at this point, we determined that we haven't seen the trace yet, so, record the
+	// traceID in the map and the spans to the storage
+
+	// calculates the item in the ring that we'll store the trace
+	sp.index = (sp.index + 1) % sp.config.NumTraces
+
+	// see if the ring has an item already
+	idAtRingItem := sp.ids[sp.index]
+	if idAtRingItem != nil {
+		// delete from the map and erase its memory entry
+		delete(sp.idToIndex, idAtRingItem.String())
+		sp.ids[sp.index] = nil
+
+		// delete from the storage
+		sp.removed <- idAtRingItem
+
+		// TODO: do we want another channel that receives evicted items? record a metric perhaps?
+		sp.logger.Debug("trace evicted", zap.Stringer("traceID", idAtRingItem))
+	}
+
+	// place the traceID in memory
+	sp.ids[sp.index] = traceID
+	sp.idToIndex[sTraceID] = sp.index
+
+	// we have the traceID in the memory, place the spans in the storage too
+	if err := sp.addSpans(traceID, trace); err != nil {
+		return fmt.Errorf("couldn't add spans to new trace: %w", err)
+	}
+
+	sp.logger.Debug("scheduled to release trace", zap.Duration("duration", sp.config.WaitDuration))
+
+	time.AfterFunc(sp.config.WaitDuration, func() {
+		sp.cancelLock.RLock()
+		if sp.cancelReleases {
+			sp.logger.Debug("processor is stopping, canceling the release of the trace", zap.Stringer("traceID", traceID))
+			return
+		}
+
+		sp.logger.Debug("releasing trace", zap.Stringer("traceID", traceID))
+		sp.expired <- traceID
+		sp.cancelLock.RUnlock()
+	})
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) processExpired(traceID pdata.TraceID) error {
+	sTraceID := traceID.String()
+	sp.logger.Debug("processing expired", zap.String("traceID", sTraceID))
+	indexToRemove, found := sp.idToIndex[sTraceID]
+
+	if !found {
+		// we likely received multiple batches with spans for the same trace
+		// and released this trace already
+		sp.logger.Debug("skipping the processing of expired trace", zap.Stringer("traceID", traceID))
+		return nil
+	}
+
+	// delete from the map and erase its memory entry
+	delete(sp.idToIndex, sTraceID)
+	sp.ids[indexToRemove] = nil
+
+	// this might block, but we don't need to wait
+	sp.logger.Debug("marking the trace as released", zap.Stringer("traceID", traceID))
+	go sp.markAsReleased(traceID)
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) markAsReleased(traceID pdata.TraceID) error {
+	// #get is a potentially blocking operation
+	trace, err := sp.st.get(traceID)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve trace %q from the storage: %w", traceID, err)
+	}
+
+	if trace.IsNil() {
+		return fmt.Errorf("the trace %q couldn't be found at the storage", traceID)
+	}
+
+	// signal that the trace is ready to be released
+	sp.logger.Debug("trace marked as released", zap.Stringer("traceID", traceID))
+	sp.released <- trace
```

Comment:
Shouldn't you also remove trace from the storage here?
```go
sp.removed <- traceID
```
This function gets called from `processExpired`, where TraceID is removed from `idToIndex` and `ids`, so we lost the pointer to item in the storage already, and this is our last chance to collect the garbage AFAICT.

Reply:
The storage has a copy of the item, it isn't a reference to the same object.

Reply:
Perhaps I wasn't clear. I didn't literally mean GC in Go.

My concern is there is no invocation of `sp.st.delete(traceID)` in this code flow.
The only way to trigger deletion of trace from storage at the moment is to push traceID to the `removed` channel, otherwise it will be stored "forever" (or far longer than needed).

The only place where you do `sp.removed <- idAtRingItem` is in the `processTrace` function, when the ring is overflown, which normally should never happen.

Reply:
Looks like I missed your last comment while editing my previous one. You are right, and the code has been fixed :-)

Reply:
Thanks for fixing this!
I still think that you need to introduce some abstractions to simplify the code and make it easier to catch this sort of errors. 

Please consider extracting the "ring buffer" into a separate struct and introducing an "event queue" with just one channel instead of 4.